### PR TITLE
CSVS-31 Recursive deletion for vault's secrets was implemented

### PR DESCRIPTION
**Pull request by MedvedevBW**

### CODE REVIEW

File: src/main/scala/com/bwsw/cloudstack/vault/server/controllers/CloudStackVaultController.scala [hunk 1]
```diff
@@ -227,7 +227,7 @@ class CloudStackVaultController(vaultService: VaultService,
       }
       zooKeeperService.deleteNode(pathToEntityNode)
     }
-    vaultService.deleteSecret(defaultSecretPath)
+    vaultService.deleteSecretsRecursive(secretPath)
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/controllers/CloudStackVaultController.scala]: rename to 'deleteSecretsRecursively'

File: src/test/scala/com/bwsw/cloudstack/vault/server/vault/util/VaultRestRequestCreatorTestSuite.scala [hunk 2]
```diff
@@ -225,7 +252,7 @@ class VaultRestRequestCreatorTestSuite extends FlatSpec with TestData with BaseT
     }
   }
 
-  "createTokenRevokeRequest" should  "if Rest throw RestException, exception does not catch" in {
+  "createTokenLookupRequest" should  "if Rest throw RestException, exception does not catch" in {
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/vault/server/vault/util/VaultRestRequestCreatorTestSuite.scala]: "createTokenLookupRequest" should  "if Rest throws RestException, the exception will not be caught"

File: src/test/scala/com/bwsw/cloudstack/vault/server/vault/VaultServiceTestSuite.scala [hunk 3]
```diff
@@ -132,19 +187,25 @@ class VaultServiceTestSuite extends FlatSpec with TestData with BaseTestSuite {
     }
   }
 
-  "deleteSecret" should "The VaultCriticalException thrown by VaultRestRequestCreator must not be swallowed" in {
-    val path = "test/path"
+  "deleteSecretRecursive" should "The VaultCriticalException thrown by VaultRestRequestCreator must not be swallowed" in {
+    val path = "/test/path"
+    val responseWithEmptySubThree = "{\"errors\":[]}"
```
Comment (placccebo) [inline src/test/scala/com/bwsw/cloudstack/vault/server/vault/VaultServiceTestSuite.scala]: fix a typo

File: src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala [hunk 4]
```diff
@@ -106,20 +107,52 @@ class VaultService(vaultRest: VaultRestRequestCreator,
...
-    TaskRunner.tryRunUntilSuccess[String](
-      executeRequest,
-      settings.vaultRetryDelay
-    )
+    def getPathListPair(pathToSecret: String): (List[String], List[String]) = {
+      jsonSerializer.deserialize[SecretResponse](TaskRunner.tryRunUntilSuccess[String](
+        vaultRest.createGetSubSecretPathsRequest(pathToSecret),
+        settings.vaultRetryDelay
+      )).secretList.getOrElse(SecretList(List.empty[String])).secrets.partition { x =>
+        Pattern.compile(".+/").matcher(x).matches()
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala]: move a pattern compilation above to not compile it for each path

File: src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala [hunk 5]
```diff
@@ -106,20 +107,52 @@ class VaultService(vaultRest: VaultRestRequestCreator,
...
+      }
+    }
 
-    logger.debug(s"data from path: $pathToSecret was deleted")
+    val subPathsOfRootPath = {
+      getPathListPair(pathToRootSecret) match {
+        case (pathsWithSubPaths, pathsWithData) =>
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala:146]: what should be done in the default case?

File: src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala [hunk 6]
```diff
@@ -106,20 +107,52 @@ class VaultService(vaultRest: VaultRestRequestCreator,
...
-  def deleteSecret(pathToSecret: String): Unit = {
-    logger.debug(s"deleteSecret: $pathToSecret")
-    def executeRequest = vaultRest.createDeleteSecretRequest(pathToSecret)
+  def deleteSecretsRecursive(pathToRootSecret: String): Unit = {
+    logger.debug(s"deleteSecretsRecursive: $pathToRootSecret")
+    var pathsForDeletion = List(pathToRootSecret)
+
+    def loop(pathToSecret: String, pathsWithSubPaths: List[String]): Unit = {
+      if (pathsWithSubPaths.nonEmpty) {
+        getPathListPair(s"$pathToSecret/${pathsWithSubPaths.head.substring(0, pathsWithSubPaths.head.length - 1)}") match {
```
Comment (placccebo) [inline src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala]: use dropRight(1) instead of substring() method

**Review by placccebo — CHANGES_REQUESTED** [review]

File: src/main/scala/com/bwsw/cloudstack/vault/server/controllers/CloudStackVaultController.scala (hunk 1 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/controllers/CloudStackVaultController.scala]: ok

File: src/test/scala/com/bwsw/cloudstack/vault/server/vault/util/VaultRestRequestCreatorTestSuite.scala (hunk 2 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/vault/server/vault/util/VaultRestRequestCreatorTestSuite.scala]: ok

File: src/test/scala/com/bwsw/cloudstack/vault/server/vault/VaultServiceTestSuite.scala (hunk 3 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/cloudstack/vault/server/vault/VaultServiceTestSuite.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala (hunk 6 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala (hunk 4 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala]: ok

File: src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala (hunk 5 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/cloudstack/vault/server/vault/VaultService.scala:146]: Fixed, though "case _" is unreachable

**Review by placccebo — APPROVED** [review]

**Pull request merged**